### DISCUSSION TITLE
Keycloak 17.0+ Best practices for theme customization

### DISCUSSION THREAD

**Discussion by ederparaiso** [keycloak/keycloak#11638] [Q&A] [answered]
Hi, I’m new into keycloak and I have some difficult to understand the development flow in order to customize keycloak. My use case is: I’ll distribute keycloak through docker images and I need to provide a whole new custom look and feel of user sign up and sign in flows.

question 1: in order to distribute keycloak as a container, I really need to build my own image from keycloak image and build it with my custom assets? For each new version of my custom assets (in this case I’m not considering only themes) I’ll need to generate a new keycloak custom image to deploy in my environments? Are there no options to avoid to maintain custom keycloak images?

question 2, about theme customization options and best practices: if I need to create a whole new html pages I need to package them as a jar file, deploy into /providers directory and build a new keycloack image? Could I have a separate frontend application calling keycloak REST APIs? (I’m not sure if Admin REST APIs have all endpoints needed).

What are your advices/best practices recommendations?

Thanks in advance

**Comment by dasniko** [comment, marked as answer]
This is a crosspost from here: https://keycloak.discourse.group/t/keycloak-17-0-best-practices-for-theme-customization/15134
